### DISCUSSION TITLE
Analyzing the gc log file is slow when the size is above 50 mb

### DISCUSSION THREAD

**Discussion by logesh-encipherhealth** [microsoft/gctoolkit#293] [General]
I use a toolkit to analyze the GC log file it takes more than 50 secs on a Linux machine and takes more than 25 secs on Mac. I gave a zip log file the original size of the log file is 90 mb after compressed it gets 9 mb. Why it takes so much time?

**Comment by karianna** [comment]
How much memory are you giving GCToolkit when you run it?

**Reply by logesh-encipherhealth** [reply]
32 gb ram  and i gave the maximum heap size of 8 GB

**Reply by karianna** [reply]
OK, and do you see any warnings in the warnings.log file?  It can take a lot of time to process the log lines but if we are finding lines that we can't parse that takes even longer (but we log that fact)

**Reply by logesh-encipherhealth** [reply]
only 2 warnings came

**Comment by dsgrieve** [comment]
@logesh056 The bulk of the time is spent in doing regular expression matching. There has been some effort to make this more efficient. The issue for this is #203

**Reply by logesh-encipherhealth** [reply]
yes correct each rule take 1 ms

**Reply by logesh-encipherhealth** [reply]
can you optimize this? it makes application to much slower

**Reply by dsgrieve** [reply]
Optimization is the purpose of issue #203. There have been some exploratory efforts at optimization. 

If you are curious about the challenges, com.microsoft.gctoolkit.parser.UnifiedG1GCParser#parse is the focal point. The method com.microsoft.gctoolkit.parser.GCParseRule#parse is where the patterns are tried. In the worse case, every pattern will be tried for a match. Attempts at optimization, such as trying the most often used patterns first, did not result in the gains we hoped for but it is not clear why.

**Reply by logesh-encipherhealth** [reply]
Okey 👍

**Reply by dsgrieve** [reply]
Hey, @logesh056. Take PR #295 for a spin. See if it helps any.

**Reply by logesh-encipherhealth** [reply]
sure @dsgrieve

**Reply by logesh-encipherhealth** [reply]
Hey @dsgrieve can you use hyperScan to find the ruleToApply?
It stores the expression in memory and it validates parseRule fast.

**Reply by dsgrieve** [reply]
I can look into it.
